src/zenml/integrations/pigeon/annotators/pigeon_annotator.py

```diff
@@ -0,0 +1,322 @@
+the original package (and code) is no longer compatible with more recent
+versions of ipywidgets.
+
+https://github.com/agermanidis/pigeon
+"""
+
+import json
+import os
+from datetime import datetime
+from typing import Any, List, Optional, Tuple, cast
+
+import ipywidgets as widgets  # type: ignore
+from IPython.core.display_functions import clear_output, display
+
+from zenml.annotators.base_annotator import BaseAnnotator
+from zenml.integrations.pigeon.flavors.pigeon_annotator_flavor import (
+    PigeonAnnotatorConfig,
+)
+from zenml.logger import get_logger
+
+logger = get_logger(__name__)
+
+
+class PigeonAnnotator(BaseAnnotator):
+    """Annotator for using Pigeon in Jupyter notebooks."""
+
+    @property
+    def config(self) -> PigeonAnnotatorConfig:
+        """Get the Pigeon annotator config.
+
+        Returns:
+            The Pigeon annotator config.
+        """
+        return cast(PigeonAnnotatorConfig, self._config)
+
+    def get_url(self) -> str:
+        """Get the URL of the Pigeon annotator.
+
+        Raises:
+            NotImplementedError: Pigeon annotator does not have a URL.
+        """
+        raise NotImplementedError("Pigeon annotator does not have a URL.")
+
+    def get_url_for_dataset(self, dataset_name: str) -> str:
+        """Get the URL of the Pigeon annotator for a specific dataset.
+
+        Args:
+            dataset_name: Name of the dataset (annotation file).
+
+        Raises:
+            NotImplementedError: Pigeon annotator does not have a URL.
+        """
+        raise NotImplementedError("Pigeon annotator does not have a URL.")
+
+    def get_datasets(self) -> List[str]:
+        """Get a list of datasets (annotation files) in the output directory.
+
+        Returns:
+            A list of dataset names (annotation file names) (or empty list when no datasets are present).
+        """
+        output_dir = self.config.output_dir
+        try:
+            return [f for f in os.listdir(output_dir) if f.endswith(".txt")]
+        except FileNotFoundError:
+            return []
+
+    def get_dataset_names(self) -> List[str]:
+        """List dataset names (annotation file names) in the output directory.
+
+        Returns:
+            A list of dataset names (annotation file names).
+        """
+        return self.get_datasets()
+
+    def get_dataset_stats(self, dataset_name: str) -> Tuple[int, int]:
+        """List labeled and unlabeled examples in a dataset (annotation file).
+
+        Args:
+            dataset_name: Name of the dataset (annotation file).
+
+        Returns:
+            A tuple containing (num_labeled_examples, num_unlabeled_examples).
+        """
+        dataset_path = os.path.join(self.config.output_dir, dataset_name)
+        num_labeled_examples = sum(1 for _ in open(dataset_path))
+        num_unlabeled_examples = 0  # Assuming all examples are labeled
+        return num_labeled_examples, num_unlabeled_examples
```

Comment:
shall we catch errors in case can't be open:
```suggestion
    dataset_path = os.path.join(self.config.output_dir, dataset_name)
    try:
        with open(dataset_path, 'r') as file:
            num_labeled_examples = sum(1 for _ in file)
        num_unlabeled_examples = 0  # Update this logic if you start handling unlabeled data
        return num_labeled_examples, num_unlabeled_examples
    except FileNotFoundError:
        logger.error(f"File not found: {dataset_path}")
        return (0, 0)
```